rpc: Fix internal bug in descriptorprocesspsbt when encountering invalid signatures

b-l-u-e opened the pull request:
Fixes #32849
descriptorprocesspsbt crashes with an internal bug assertion when given a PSBT whose inputs looked “finalized” that is non-empty final_script_sig / witness but whose signatures did not verify like invalid Schnorr bytes with unusual sighash combinations such as SIGHASH_SINGLE | ANYONECANPAY.

Completeness was inferred with PSBTInputSigned, which only checks that final fields are present, not that they pass script verification. That let the RPC treat the PSBT as complete and call FinalizeAndExtractPSBT, where a failing invariant surfaced as CHECK_NONFATAL instead of a normal incomplete outcome.

This PR determines complete the same way as the wallet path: after ProcessPSBT, it recomputes PrecomputedTransactionData and sets complete only if every input passes PSBTInputSignedAndVerified. 
Invalid signatures then yield complete: false and no hex, without going through the finalize path that asserted.

achow101 commented:
The `CHECK_NONFATAL` is correct. `complete == true` means that the PSBT is complete and can be finalized. The invariant here is that `complete == true` means `FinalizeAndExtract` must succeed.

BIP 174 states that a finalizer must only construct valid scriptSigs and witnesses. We encounter this error because we assume that if a PSBT is finalized, the scriptSigs and witnesses are valid so we are not performing the extra step of verifying the script. However, in this situation, the finalizer has misbehaved and produce invalid scriptSigs or witnesses which causes the assumption to fail.

This should be resolved by using `PSBTInputSignedAndVerified` instead. It may even be worth it to combine `PSBTInputSigned` with `PSBTInputSignedAndVerified` and replacing `PSBTInputSigned` wherever it is used. Although that may not be possible to do.

b-l-u-e commented:
@achow101...You were right, my initial approach was incorrect. I now understand that CHECK_NONFATAL is correctly protecting the invariant that a PSBT marked as complete must be finalizable.

The actual bug, as you pointed out, was that the completeness check was too lenient, only checking for the presence of a signature rather than its cryptographic validity.

After ensuring that the logic first verifies the existing signatures before declaring the PSBT complete (the approach you suggested with PSBTInputSignedAndVerified), the system now behaves as expected. The same invalid PSBT now correctly results in { "complete": false } instead of triggering the internal bug.

rkrux reviewed (commented):
I don't think the corresponding bug is still present after v30, see: https://github.com/bitcoin/bitcoin/issues/32849#issuecomment-3443211089

Also, a test case in `rpc_psbt.py` would be good to verify the fix.

l0rinc commented on src/rpc/rawtransaction.cpp:
```diff
@@ -2088,8 +2088,9 @@ RPCHelpMan descriptorprocesspsbt()
-        complete &= PSBTInputSigned(input);
+    const PrecomputedTransactionData txdata = PrecomputePSBTData(psbtx);
+    for (unsigned int i = 0; i < psbtx.inputs.size(); ++i) {
+        complete &= PSBTInputSignedAndVerified(psbtx, i, &txdata);
```
this is likely a change in behavior since && short circuits

rkrux reviewed (commented):
Concept ACK 507501ddfa3f2c538a0817a5645b20ddde554362

> I don't think the corresponding bug is still present after v30, see: https://github.com/bitcoin/bitcoin/issues/32849#issuecomment-3443211089

I was able to reproduce this bug in the functional tests.

> Also, a test case in rpc_psbt.py would be good to verify the fix.

Consider adding the following test case in the second commit in this PR. The following test fails without this fix and passes with it.

**Edit:** Added `bitflipper` function in util (taken from [feature_taproot.py](https://github.com/bitcoin/bitcoin/blob/f54ffb4bc141ffcc6fffe574edbcbbb21f001476/test/functional/feature_taproot.py#L606-L612)) and used it to make the signature invalid. The previous approach I used here changed the last byte to `0x42` always that could have been proven to be flaky.

```diff
diff --git a/test/functional/rpc_psbt.py b/test/functional/rpc_psbt.py
index 933e4cd1ad..55c5df4a95 100755
--- a/test/functional/rpc_psbt.py
+++ b/test/functional/rpc_psbt.py
@@ -37,6 +37,7 @@ from test_framework.psbt import (
     PSBT_IN_WITNESS_UTXO,
     PSBT_OUT_MUSIG2_PARTICIPANT_PUBKEYS,
     PSBT_OUT_TAP_TREE,
+    PSBT_IN_FINAL_SCRIPTWITNESS,
 )
 from test_framework.script import CScript, OP_TRUE, SIGHASH_ALL, SIGHASH_ANYONECANPAY
 from test_framework.script_util import MIN_STANDARD_TX_NONWITNESS_SIZE
@@ -50,6 +51,7 @@ from test_framework.util import (
     assert_raises_rpc_error,
     find_vout_for_address,
     wallet_importprivkey,
+    bitflipper,
 )
 from test_framework.wallet_util import (
     calculate_input_weight,
@@ -369,6 +371,35 @@ class PSBTTest(BitcoinTestFramework):
         changepos = psbtx["changepos"]
         assert_equal(decoded_psbt["tx"]["vout"][changepos]["scriptPubKey"]["type"], expected_type)
 
+    def test_psbt_with_invalid_signature(self):
+        self.log.info("Test descriptorprocesspsbt with invalid signature in signed PSBT")
+
+        def_wallet = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+        outputs = [{def_wallet.getnewaddress(address_type="bech32m"): 1}]
+        def_wallet.send(outputs)
+        self.generate(self.nodes[0], 6)
+
+        utxos = [utxo for utxo in def_wallet.listunspent() if utxo["desc"].startswith("tr")]
+        psbt = def_wallet.walletcreatefundedpsbt(utxos, [{def_wallet.getnewaddress(): 0.5}])["psbt"]
+        descs = [desc for desc in def_wallet.listdescriptors(True)["descriptors"] if desc["desc"].startswith("tr")]
+
+        # unload wallet so as to not use the wallet RPC in further process PSBT calls
+        def_wallet.unloadwallet()
+
+        desc_psbt = self.nodes[0].descriptorprocesspsbt(psbt=psbt, descriptors=descs, finalize=True)["psbt"]
+        flawed_psbt = PSBT.from_base64(desc_psbt)
+        valid_sig = flawed_psbt.i[0].map.get(PSBT_IN_FINAL_SCRIPTWITNESS)
+        invalid_sig = bitflipper(valid_sig)
+        flawed_psbt.i[0].map[PSBT_IN_FINAL_SCRIPTWITNESS] = invalid_sig
+        flawed_psbt = flawed_psbt.to_base64()
+
+        result = self.nodes[0].descriptorprocesspsbt(psbt=flawed_psbt, descriptors=descs)
+        assert_equal(result["complete"], False)
+        assert_equal("hex" in result, False)
+
+        # load the default wallet back
+        self.nodes[0].loadwallet(self.default_wallet_name)
+
     def run_test(self):
         # Create and fund a raw tx for sending 10 BTC
         psbtx1 = self.nodes[0].walletcreatefundedpsbt([], {self.nodes[2].getnewaddress():10})['psbt']
@@ -1211,6 +1242,7 @@ class PSBTTest(BitcoinTestFramework):
         if not self.options.usecli:
             self.test_sighash_mismatch()
         self.test_sighash_adding()
+        self.test_psbt_with_invalid_signature()
 
 if __name__ == '__main__':
     PSBTTest(__file__).main()
diff --git a/test/functional/test_framework/util.py b/test/functional/test_framework/util.py
index bbad6ea771..b4e597ce1d 100644
--- a/test/functional/test_framework/util.py
+++ b/test/functional/test_framework/util.py
@@ -715,3 +715,6 @@ def wallet_importprivkey(wallet_rpc, privkey, timestamp, *, label=""):
     }]
     import_res = wallet_rpc.importdescriptors(req)
     assert_equal(import_res[0]["success"], True)
+
+def bitflipper(input):
+    assert isinstance(input, bytes)
+    return (int.from_bytes(input, 'little') ^ (1 << random.randrange(len(input) * 8))).to_bytes(len(input), 'little')
\ No newline at end of file

```

achow101 commented:
A test is definitely necessary here to avoid future regressions. @rkrux's proposed test would be fine.

The commit message has a duplicate line that needs to be removed.

rkrux reviewed (approved):
lgtm ACK ac99af4f80191caa24badcfbb5cca6811fc92cf1

Thanks for accepting the suggestion adding the functional test.

rkrux reviewed (approved):
code review re-ACK 7e468a6

```diff
git range-diff ac99af4...7e468a6
```

Thanks for accepting the nit suggestion.

achow101 commented on test/functional/rpc_psbt.py:
```diff
@@ -406,6 +408,39 @@ def test_psbt_named_parameter_handling(self):
+
+        # unload wallet so as to not use the wallet RPC in further process PSBT calls
+        def_wallet.unloadwallet()
+        
```
In 7e468a6fb01b36edb6955ccb2b869ec1063a891c "rpc: Fix descriptorprocesspsbt internal bug on invalid signatures"

Extra whitespace on this line

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -2048,10 +2048,12 @@ RPCHelpMan descriptorprocesspsbt()
         finalize);
 
-    // Check whether or not all of the inputs are now signed
+    // Check whether or not all of the inputs are now correctly signed 
```
In 7e468a6fb01b36edb6955ccb2b869ec1063a891c "rpc: Fix descriptorprocesspsbt internal bug on invalid signatures"

Trailing whitespace on this line

maflcko commented on test/functional/rpc_psbt.py:
```diff
@@ -136,7 +138,7 @@ def test_utxo_conversion(self):
         # Have the offline node sign the PSBT (which will remove the non-witness UTXO)
         signed_psbt = offline_node.walletprocesspsbt(psbt_new.to_base64())
-        assert not "non_witness_utxo" in mining_node.decodepsbt(signed_psbt["psbt"])["inputs"][0]
+        assert "non_witness_utxo" not in mining_node.decodepsbt(signed_psbt["psbt"])["inputs"][0]
```
i guess it could make sense to enable this globally, so i've done that in https://github.com/bitcoin/bitcoin/pull/34154. Mixing unrelated refactoring cleanup with a bugfix in the same commit often makes review harder.

rkrux reviewed (approved):
re-ACK 9f55be650e6181f7d3ea898e3ccfafedecc26869

Removed the unrelated lint changes because PR #34154 is merged now.
```zsh
git range-diff 222139e...9f55be6
```

Please resolve the PR comments that have been addressed to make it easier to go through the PR.

b-l-u-e commented:
> re-ACK [9f55be6](https://github.com/bitcoin/bitcoin/commit/9f55be650e6181f7d3ea898e3ccfafedecc26869)
> 
> Removed the unrelated lint changes because PR #34154 is merged now.
> 
> ```shell
> git range-diff 222139e...9f55be6
> ```
> 
> Please resolve the PR comments that have been addressed to make it easier to go through the PR.

all the PR comments have been resolved

b-l-u-e commented:
> ```
> git range-diff 9f55be6...e995c85
> ```
> 
> I don't see any changes in the latest range-diff, why was this latest push done after this [#33014 (review)](https://github.com/bitcoin/bitcoin/pull/33014#pullrequestreview-3626721928)?

the review my bad i was suppose to comment after 
i rebased the branch

achow101 reviewed (commented):
ACK 5a04a06f1f782b7d19a35b44cc8fe81f2b4d4c55

rkrux reviewed (commented):
+1 on fixing the error https://github.com/bitcoin/bitcoin/pull/33014#discussion_r2663883608 highlighted by Drahbot, I didn't think of this scenario earlier.

rkrux reviewed (approved):
ACK 2fd79d8
Thanks for following up on all the previous comments.

```
git range-diff 5a04a06...2fd79d8
```

Commit:
rpc: Fix descriptorprocesspsbt internal bug on invalid signatures

Use PSBTInputSignedAndVerified instead of PSBTInputSigned to properly
validate signatures before marking PSBT as complete

Signed-off-by: b-l-u-e <8102260+blue@users.noreply.github.com>

rkrux commented on src/rpc/rawtransaction.cpp:
```diff
@@ -2112,10 +2112,16 @@ RPCMethod descriptorprocesspsbt()
+    std::optional<PrecomputedTransactionData> txdata_res = PrecomputePSBTData(psbtx);
+    if (!txdata_res) {
+        throw JSONRPCPSBTError(common::PSBTError::INVALID_TX);
+    }
```
This check already happens in ProcessPSBT call above, it need not be done again and instead can be replaced with an assert.

https://github.com/bitcoin/bitcoin/blob/3f9c55426a12a16bd11de6dec139c4df58788360/src/rpc/rawtransaction.cpp#L183-L186

```diff
diff --git a/src/rpc/rawtransaction.cpp b/src/rpc/rawtransaction.cpp
index a90614e334..6b218da66d 100644
--- a/src/rpc/rawtransaction.cpp
+++ b/src/rpc/rawtransaction.cpp
@@ -2115,9 +2115,7 @@ RPCMethod descriptorprocesspsbt()
     // Check whether or not all of the inputs are now correctly signed
     bool complete = true;
     std::optional<PrecomputedTransactionData> txdata_res = PrecomputePSBTData(psbtx);
-    if (!txdata_res) {
-        throw JSONRPCPSBTError(common::PSBTError::INVALID_TX);
-    }
+    Assert(txdata_res.has_value());
     const PrecomputedTransactionData& txdata = *txdata_res;
     for (unsigned int i = 0; i < psbtx.inputs.size(); ++i) {
         const bool verified = PSBTInputSignedAndVerified(psbtx, i, &txdata);

```

b-l-u-e replied:
thank you for suggestion made change

b-l-u-e replied:
i faced an error
```
src/rpc/rawtransaction.cpp:2118:    Assert(txdata_res.has_value());
^^^
CHECK_NONFATAL(condition) or NONFATAL_UNREACHABLE should be used instead of assert for RPC code.

Aborting the whole process is undesirable for RPC code. So nonfatal
checks should be used over assert. See: src/util/check.h
^---- ⚠️ Failure generated from lint check 'rpc_assert' (Check that fatal assertions are not used in RPC code)!
```

replaced `Assert(txdata_res.has_value());` with `CHECK_NONFATAL(txdata_res.has_value());`

rkrux reviewed (changes requested):
Code review at eb6fb09419399bad0f704b4ada9b7bfe738f1429

In the PR description, the following is outdated. Please update it.

> This change replaces the CHECK_NONFATAL call with proper error handling that catches the failure and returns a JSONRPCError with a clear message to the user.

maflcko commented on src/rpc/rawtransaction.cpp:
```diff
@@ -2112,10 +2112,13 @@ RPCMethod descriptorprocesspsbt()
-        complete &= PSBTInputSigned(input);
+    std::optional<PrecomputedTransactionData> txdata_res = PrecomputePSBTData(psbtx);
+    CHECK_NONFATAL(txdata_res.has_value());
+    const PrecomputedTransactionData& txdata = *txdata_res;
```
nit: Those 3 lines can be a single one: `const PrecomputedTransactionData txdata{*CHECK_NONFATAL(PrecomputePSBTData(psbtx));`

rkrux replied:
This suggestion fails to compile in my system.

```zsh
/Users/rkrux/projects/bitcoin/src/rpc/rawtransaction.cpp:2117:11: error: unknown type name '_'; did you mean 'mpl_::_'?
 2117 |     const _& txdata{*CHECK_NONFATAL(PrecomputePSBTData(psbtx));
      |           ^
      |           mpl_::_
/usr/local/Cellar/boost/1.90.0_1/include/boost/mpl/aux_/preprocessed/gcc/placeholders.hpp:14:19: note: 'mpl_::_' declared here
   14 | typedef arg< -1 > _;
      |                   ^
/Users/rkrux/projects/bitcoin/src/rpc/rawtransaction.cpp:2117:63: error: expected '}'
 2117 |     const _& txdata{*CHECK_NONFATAL(PrecomputePSBTData(psbtx));
      |                                                               ^
/Users/rkrux/projects/bitcoin/src/rpc/rawtransaction.cpp:2117:20: note: to match this '{'
 2117 |     const _& txdata{*CHECK_NONFATAL(PrecomputePSBTData(psbtx));
      |                    ^
/Users/rkrux/projects/bitcoin/src/rpc/rawtransaction.cpp:2119:32: error: no matching function for call to 'PSBTInputSignedAndVerified'
 2119 |         complete = complete && PSBTInputSignedAndVerified(psbtx, i, &txdata);
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/rkrux/projects/bitcoin/src/psbt.h:1641:6: note: candidate function not viable: no known conversion from 'const mpl_::_ *' (aka 'const arg<-1> *') to 'const PrecomputedTransactionData *' for 3rd argument
 1641 | bool PSBTInputSignedAndVerified(const PartiallySignedTransaction& psbt, unsigned int input_index, const PrecomputedTransactionData* txdata);
      |      ^                                                                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 errors generated.
make[2]: *** [src/CMakeFiles/bitcoin_node.dir/rpc/rawtransaction.cpp.o] Error 1
```

maflcko replied:
I used `_` as a placeholder for the type `PrecomputedTransactionData`, because I thought it would be obvious, but I've edited my comment now. :sweat_smile:

rkrux replied:
> I used _ as a placeholder for the type

Oh, I didn't know it was a thing.

maflcko replied:
Ah, also the `&` isn't needed

b-l-u-e replied:
> nit: Those 3 lines can be a single one: `const PrecomputedTransactionData txdata{*CHECK_NONFATAL(PrecomputePSBTData(psbtx));`

Thanks for the nit. I tried the one-liner but on  CI the build fails 
```
 2117 |     const PrecomputedTransactionData& txdata{*CHECK_NONFATAL(PrecomputePSBTData(psbtx))};
      |                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/admin/actions-runner/_work/_temp/src/util/check.h:113:28: note: expanded from macro 'CHECK_NONFATAL'
  113 |     inline_check_non_fatal(condition, std::source_location::current(), #condition)
      |                            ^~~~~~~~~
1 error generated.
gmake[2]: *** [src/CMakeFiles/bitcoin_node.dir/build.make:1182: src/CMakeFiles/bitcoin_node.dir/rpc/rawtransaction.cpp.o] Error 1
gmake[2]: Leaving directory '/home/admin/actions-runner/_work/_temp/build'
gmake[1]: *** [CMakeFiles/Makefile2:580: src/CMakeFiles/bitcoin_node.dir/all] Error 2
gmake[1]: Leaving directory '/home/admin/actions-runner/_work/_temp/build'
gmake: *** [Makefile:91: all] Error 2
```

i splitted into two instead first store the result in txdata_opt then use *CHECK_NONFATAL(txdata_opt)

rkrux reviewed (approved):
lgtm re-ACK 7e19ce200b2e65770907a818b02e4ec3da9c5374

Hope this PR gets merged soon.

sedited commented:
@achow101 can you take another look?

achow101 commented:
ACK 7e19ce200b2e65770907a818b02e4ec3da9c5374
